File: go/test/cls_with_options_test.go
package test

import (
	"testing"
	"time"

	modal "github.com/modal-labs/modal-client/go"
	"github.com/modal-labs/modal-client/go/internal/grpcmock"
	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"github.com/onsi/gomega"
)

var mockFunctionProto = pb.FunctionGetResponse_builder{
	FunctionId: "fid",
	HandleMetadata: pb.FunctionHandleMetadata_builder{
		MethodHandleMetadata: map[string]*pb.FunctionHandleMetadata{"echo_string": {}},
		ClassParameterInfo:   pb.ClassParameterInfo_builder{Schema: []*pb.ClassParameterSpec{}}.Build(),
	}.Build(),
}.Build()

func TestClsWithOptionsStacking(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	grpcmock.HandleUnary(
		mock, "FunctionBindParams",
		func(req *pb.FunctionBindParamsRequest) (*pb.FunctionBindParamsResponse, error) {
			g.Expect(req.GetFunctionId()).To(gomega.Equal("fid"))
			fo := req.GetFunctionOptions()
			g.Expect(fo).ToNot(gomega.BeNil())
			g.Expect(fo.GetTimeoutSecs()).To(gomega.Equal(uint32(60)))
			g.Expect(fo.GetResources()).ToNot(gomega.BeNil())
			g.Expect(fo.GetResources().GetMilliCpu()).To(gomega.Equal(uint32(250)))
			g.Expect(fo.GetResources().GetMemoryMb()).To(gomega.Equal(uint32(256)))
			g.Expect(fo.GetResources().GetGpuConfig()).ToNot(gomega.BeNil())
			g.Expect(fo.GetSecretIds()).To(gomega.Equal([]string{"sec-1"}))
			g.Expect(fo.GetReplaceSecretIds()).To(gomega.BeTrue())
			g.Expect(fo.GetReplaceVolumeMounts()).To(gomega.BeTrue())
			g.Expect(fo.GetVolumeMounts()).To(gomega.HaveLen(1))
			g.Expect(fo.GetVolumeMounts()[0].GetMountPath()).To(gomega.Equal("/mnt/test"))
			g.Expect(fo.GetVolumeMounts()[0].GetVolumeId()).To(gomega.Equal("vol-1"))
			g.Expect(fo.GetVolumeMounts()[0].GetAllowBackgroundCommits()).To(gomega.BeTrue())
			g.Expect(fo.GetVolumeMounts()[0].GetReadOnly()).To(gomega.BeFalse())
			return pb.FunctionBindParamsResponse_builder{BoundFunctionId: "fid-1", HandleMetadata: &pb.FunctionHandleMetadata{}}.Build(), nil
		},
	)

	secret := &modal.Secret{SecretID: "sec-1"}
	volume := &modal.Volume{VolumeID: "vol-1"}
	cpu := 0.25
	memoryMiB := 256
	gpu := "T4"
	timeout := 45 * time.Second
	newTimeout := 60 * time.Second

	optioned := cls.
		WithOptions(&modal.ClsWithOptionsParams{Timeout: &timeout, CPU: &cpu}).
		WithOptions(&modal.ClsWithOptionsParams{Timeout: &newTimeout, MemoryMiB: &memoryMiB, GPU: &gpu}).
		WithOptions(&modal.ClsWithOptionsParams{Secrets: []*modal.Secret{secret}, Volumes: map[string]*modal.Volume{"/mnt/test": volume}})

	instance, err := optioned.Instance(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(instance).ToNot(gomega.BeNil())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestClsWithConcurrencyWithBatchingChaining(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	grpcmock.HandleUnary(
		mock, "FunctionBindParams",
		func(req *pb.FunctionBindParamsRequest) (*pb.FunctionBindParamsResponse, error) {
			g.Expect(req.GetFunctionId()).To(gomega.Equal("fid"))
			fo := req.GetFunctionOptions()
			g.Expect(fo).ToNot(gomega.BeNil())
			g.Expect(fo.GetTimeoutSecs()).To(gomega.Equal(uint32(60)))
			g.Expect(fo.GetMaxConcurrentInputs()).To(gomega.Equal(uint32(10)))
			g.Expect(fo.GetBatchMaxSize()).To(gomega.Equal(uint32(11)))
			g.Expect(fo.GetBatchLingerMs()).To(gomega.Equal(uint64(12)))
			return pb.FunctionBindParamsResponse_builder{BoundFunctionId: "fid-1", HandleMetadata: &pb.FunctionHandleMetadata{}}.Build(), nil
		},
	)

	timeout := 60 * time.Second
	chained := cls.
		WithOptions(&modal.ClsWithOptionsParams{Timeout: &timeout}).
		WithConcurrency(&modal.ClsWithConcurrencyParams{MaxInputs: 10}).
		WithBatching(&modal.ClsWithBatchingParams{MaxBatchSize: 11, Wait: 12 * time.Millisecond})

	instance, err := chained.Instance(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(instance).ToNot(gomega.BeNil())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestClsWithOptionsRetries(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	grpcmock.HandleUnary(
		mock, "FunctionBindParams",
		func(req *pb.FunctionBindParamsRequest) (*pb.FunctionBindParamsResponse, error) {
			fo := req.GetFunctionOptions()
			g.Expect(fo).ToNot(gomega.BeNil())
			g.Expect(fo.GetRetryPolicy()).ToNot(gomega.BeNil())
			g.Expect(fo.GetRetryPolicy().GetRetries()).To(gomega.Equal(uint32(2)))
			g.Expect(fo.GetRetryPolicy().GetBackoffCoefficient()).To(gomega.Equal(float32(2.0)))
			g.Expect(fo.GetRetryPolicy().GetInitialDelayMs()).To(gomega.Equal(uint32(2000)))
			g.Expect(fo.GetRetryPolicy().GetMaxDelayMs()).To(gomega.Equal(uint32(5000)))
			return pb.FunctionBindParamsResponse_builder{BoundFunctionId: "fid-2", HandleMetadata: &pb.FunctionHandleMetadata{}}.Build(), nil
		},
	)

	backoff := float32(2.0)
	initial := 2 * time.Second
	max := 5 * time.Second
	retries, err := modal.NewRetries(2, &modal.RetriesParams{
		BackoffCoefficient: &backoff,
		InitialDelay:       &initial,
		MaxDelay:           &max,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{Retries: retries}).Instance(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestClsWithOptionsInvalidValues(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	timeout := 500 * time.Millisecond
	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{Timeout: &timeout}).Instance(ctx, nil)
	g.Expect(err).Should(gomega.HaveOccurred())
	g.Expect(err.Error()).Should(gomega.ContainSubstring("timeout must be at least 1 second"))

	scaledownWindow := 100 * time.Millisecond
	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{ScaledownWindow: &scaledownWindow}).Instance(ctx, nil)
	g.Expect(err).Should(gomega.HaveOccurred())
	g.Expect(err.Error()).Should(gomega.ContainSubstring("scaledownWindow must be at least 1 second"))

	fractionalTimeout := 1500 * time.Millisecond
	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{Timeout: &fractionalTimeout}).Instance(ctx, nil)
	g.Expect(err).Should(gomega.HaveOccurred())
	g.Expect(err.Error()).Should(gomega.ContainSubstring("whole number of seconds"))

	fractionalScaledown := 1500 * time.Millisecond
	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{ScaledownWindow: &fractionalScaledown}).Instance(ctx, nil)
	g.Expect(err).Should(gomega.HaveOccurred())
	g.Expect(err.Error()).Should(gomega.ContainSubstring("whole number of seconds"))

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestWithOptionsEmptySecretsDoesNotReplace(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	grpcmock.HandleUnary(
		mock, "FunctionBindParams",
		func(req *pb.FunctionBindParamsRequest) (*pb.FunctionBindParamsResponse, error) {
			g.Expect(req.GetFunctionId()).To(gomega.Equal("fid"))
			fo := req.GetFunctionOptions()
			g.Expect(fo.GetSecretIds()).To(gomega.HaveLen(0))
			g.Expect(fo.GetReplaceSecretIds()).To(gomega.BeFalse())

			return pb.FunctionBindParamsResponse_builder{BoundFunctionId: "fid-1", HandleMetadata: &pb.FunctionHandleMetadata{}}.Build(), nil
		},
	)

	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{Secrets: []*modal.Secret{}}).Instance(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestWithOptionsEmptyVolumesDoesNotReplace(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "FunctionGet",
		func(req *pb.FunctionGetRequest) (*pb.FunctionGetResponse, error) {
			return mockFunctionProto, nil
		},
	)

	cls, err := mock.Cls.FromName(ctx, "libmodal-test-support", "EchoCls", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	grpcmock.HandleUnary(
		mock, "FunctionBindParams",
		func(req *pb.FunctionBindParamsRequest) (*pb.FunctionBindParamsResponse, error) {
			g.Expect(req.GetFunctionId()).To(gomega.Equal("fid"))
			fo := req.GetFunctionOptions()
			g.Expect(fo.GetVolumeMounts()).To(gomega.HaveLen(0))
			g.Expect(fo.GetReplaceVolumeMounts()).To(gomega.BeFalse())

			return pb.FunctionBindParamsResponse_builder{BoundFunctionId: "fid-1", HandleMetadata: &pb.FunctionHandleMetadata{}}.Build(), nil
		},
	)

	_, err = cls.WithOptions(&modal.ClsWithOptionsParams{Volumes: map[string]*modal.Volume{}}).Instance(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

File: js/test/cls_with_options.test.ts
import { expect, test } from "vitest";
import type { Secret, Volume } from "modal";
import { createMockModalClients } from "../test-support/grpc_mock";
import { Retries } from "modal";

const _mockFunctionProto = {
  functionId: "fid",
  handleMetadata: {
    methodHandleMetadata: { echo_string: {} },
    classParameterInfo: { schema: [] },
  },
};

test("Cls.withOptions stacking", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(fo.timeoutSecs).toBe(60);
    expect(fo.resources.milliCpu).toBe(250);
    expect(fo.resources.memoryMb).toBe(256);
    expect(fo.resources.gpuConfig).toBeDefined();
    expect(fo.secretIds).toEqual(["sec-1"]);
    expect(fo.replaceSecretIds).toBe(true);
    expect(fo.replaceVolumeMounts).toBe(true);
    expect(fo.volumeMounts).toEqual([
      {
        mountPath: "/mnt/test",
        volumeId: "vol-1",
        allowBackgroundCommits: true,
        readOnly: false,
      },
    ]);
    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const secret = { secretId: "sec-1" } as Secret;
  const volume = { volumeId: "vol-1" } as Volume;

  const optioned = cls
    .withOptions({ timeoutMs: 45_000, cpu: 0.25 })
    .withOptions({ timeoutMs: 60_000, memoryMiB: 256, gpu: "T4" })
    .withOptions({ secrets: [secret], volumes: { "/mnt/test": volume } });

  const instance = await optioned.instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("Cls.withConcurrency/withConcurrency/withBatching chaining", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(fo.timeoutSecs).toBe(60);
    expect(fo.maxConcurrentInputs).toBe(10);
    expect(fo.batchMaxSize).toBe(11);
    expect(fo.batchLingerMs).toBe(12);
    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const chained = cls
    .withOptions({ timeoutMs: 60_000 })
    .withConcurrency({ maxInputs: 10 })
    .withBatching({ maxBatchSize: 11, waitMs: 12 });

  const instance = await chained.instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("Cls.withOptions retries", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");

  mock.handleUnary("FunctionBindParams", (req: any) => {
    const fo = req.functionOptions;
    expect(fo.retryPolicy).toMatchObject({
      retries: 3,
      backoffCoefficient: 1.0,
      initialDelayMs: 1000,
      maxDelayMs: 60000,
    });
    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  await cls.withOptions({ retries: 3 }).instance();

  mock.handleUnary("FunctionBindParams", (req: any) => {
    const fo = req.functionOptions;
    expect(fo.retryPolicy).toMatchObject({
      retries: 2,
      backoffCoefficient: 2.0,
      initialDelayMs: 2000,
      maxDelayMs: 5000,
    });
    return { boundFunctionId: "fid-2", handleMetadata: {} };
  });

  const retries = new Retries({
    maxRetries: 2,
    backoffCoefficient: 2.0,
    initialDelayMs: 2000,
    maxDelayMs: 5000,
  });
  await cls.withOptions({ retries }).instance();

  mock.assertExhausted();
});

test("Cls.withOptions invalid values", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ timeoutMs: 1500 }).instance()).rejects.toThrow(
    /timeoutMs must be a multiple of 1000ms/,
  );

  await expect(
    cls.withOptions({ scaledownWindowMs: 2500 }).instance(),
  ).rejects.toThrow(/scaledownWindowMs must be a multiple of 1000ms/);

  mock.assertExhausted();
});

test("withOptions({ secrets: [] }) binds and does not replace secrets", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(Array.isArray(fo.secretIds)).toBe(true);
    expect(fo.secretIds.length).toBe(0);
    expect(fo.replaceSecretIds).toBe(false);

    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  const instance = await cls.withOptions({ secrets: [] }).instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("withOptions({ volumes: {} }) binds and does not replace volumes", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();
  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(Array.isArray(fo.volumeMounts)).toBe(true);
    expect(fo.volumeMounts.length).toBe(0);
    expect(fo.replaceVolumeMounts).toBe(false);

    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  const instance = await cls.withOptions({ volumes: {} }).instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("withOptions({ cpu, cpuLimit }) sets milliCpu and milliCpuMax", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(fo.resources.milliCpu).toBe(2000);
    expect(fo.resources.milliCpuMax).toBe(4500);
    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  const instance = await cls
    .withOptions({ cpu: 2.0, cpuLimit: 4.5 })
    .instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("withOptions cpuLimit lower than cpu throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(
    cls.withOptions({ cpu: 4.0, cpuLimit: 2.0 }).instance(),
  ).rejects.toThrow("cpu (4) cannot be higher than cpuLimit (2)");

  mock.assertExhausted();
});

test("withOptions cpuLimit without cpu throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ cpuLimit: 4.0 }).instance()).rejects.toThrow(
    "must also specify cpu when cpuLimit is specified",
  );

  mock.assertExhausted();
});

test("withOptions({ memory, memoryLimit }) sets memoryMb and memoryMbMax", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  mock.handleUnary("FunctionBindParams", (req: any) => {
    expect(req).toMatchObject({ functionId: "fid" });
    const fo = req.functionOptions;
    expect(fo.resources.memoryMb).toBe(1024);
    expect(fo.resources.memoryMbMax).toBe(2048);
    return { boundFunctionId: "fid-1", handleMetadata: {} };
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  const instance = await cls
    .withOptions({ memoryMiB: 1024, memoryLimitMiB: 2048 })
    .instance();
  expect(instance).toBeTruthy();

  mock.assertExhausted();
});

test("withOptions memoryLimit lower than memory throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(
    cls.withOptions({ memoryMiB: 2048, memoryLimitMiB: 1024 }).instance(),
  ).rejects.toThrow(
    "memoryMiB (2048) cannot be higher than memoryLimitMiB (1024)",
  );

  mock.assertExhausted();
});

test("withOptions memoryLimit without memory throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(
    cls.withOptions({ memoryLimitMiB: 2048 }).instance(),
  ).rejects.toThrow(
    "must also specify memoryMiB when memoryLimitMiB is specified",
  );

  mock.assertExhausted();
});

test("withOptions negative cpu throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ cpu: -1.0 }).instance()).rejects.toThrow(
    "must be a positive number",
  );

  mock.assertExhausted();
});

test("withOptions zero cpu throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ cpu: 0.0 }).instance()).rejects.toThrow(
    "must be a positive number",
  );

  mock.assertExhausted();
});

test("withOptions negative memory throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ memoryMiB: -100 }).instance()).rejects.toThrow(
    "must be a positive number",
  );

  mock.assertExhausted();
});

test("withOptions zero memory throws error", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("FunctionGet", (_: any) => {
    return _mockFunctionProto;
  });

  const cls = await mc.cls.fromName("libmodal-test-support", "EchoCls");
  await expect(cls.withOptions({ memoryMiB: 0 }).instance()).rejects.toThrow(
    "must be a positive number",
  );

  mock.assertExhausted();
});
